### ISSUE TITLE
Compiler warnings about inefficient incremental compilation

### ISSUE THREAD

**Issue by weiznich** [rust-lang/rust#40766]
While compiling [diesel](https://github.com/diesel-rs/diesel/issues/824)/[ring](https://github.com/briansmith/ring/issues/493) rustc emitted the following warning:
```
warning: Quasi-quoting might make incremental compilation very inefficient: NtExpr(..)
```

@sgrif  suspects this to be a bug in rustc rather than in diesel. So I will report this also here.

Environment:
rustc 1.17.0-nightly (8c4f2c64c 2017-03-22)
export CARGO_INCREMENTAL=1 RUSTFLAGS="-Zincremental-info"

**Comment by sgrif** [comment]
This is coming from the use of `#[doc=$macro_var]` inside of a macro it seems

**Comment by briansmith** [comment]
> This is coming from the use of `#[doc=$macro_var]` inside of a macro it seems

Yes, this seems to be what *ring* is doing that triggers this warning.

Does the use of `#[doc=$macro_var]` actually hurt incremental compilation performance? I.e. is this warning a false positive or is it a real issue when `#[doc=$macro_var]` is used?

If it really is causing performance problems, is there something I can do that is equivalent that won't? And/or is there something that can be changed in rustc to fix the compilation performance for this construct?

If it isn't really causing incremental compilation to be slow, then could we change rustc to stop emitting this warning. And/or could we add a way to use `#[allow(...)]` to turn off the warning?

**Comment by michaelwoerister** [comment]
> Does the use of #[doc=$macro_var] actually hurt incremental compilation performance?

I don't know how `#[doc=$macro_var]` works but the thing it is attached to will probably always be considered changed by incr. comp., even if it hasn't.

**Comment by briansmith** [comment]
> I don't know how #[doc=$macro_var] works but the thing it is attached to will probably always be considered changed by incr. comp., even if it hasn't.

It seems strange that that would be true, since it's just documentation, which doesn't affect code generation. IMO documentation should be ignored when comparing things for changes.

**Comment by michaelwoerister** [comment]
> It seems strange that that would be true, since it's just documentation, which doesn't affect code generation.

If we can guarantee that it has no influence on a regular compilation session then yes, we can bake special knowledge into the compiler to ignore this attribute. This must be some kind of special case though, because we already [filter out](https://github.com/rust-lang/rust/blob/5309a3e31d88def1f3ea966162ed4f81f161d500/src/librustc_incremental/calculate_svh/svh_visitor.rs#L960) doc comments.

Does someone have a minimal test case for this lying around?

**Comment by michaelwoerister** [comment]
I'm closing this in favor of https://github.com/rust-lang/rust/issues/40946 where there's been some concrete discussion about a solution.
